repositories/apiserver.go

```diff
@@ -404,6 +408,325 @@ func (a *APIServerStore) StoreCVESummary(ctx context.Context, cve domain.CVEMani
+		} else {
+			return err
+		}
+	} else {
+		// Update VEX
+		err = a.updateVEX(ctx, cve, cvep, withRelevancy, vexContainer)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+func createProductStructForImageAndPackage(imagePullable string, packagePURL string) (*v1beta1.Product, error) {
+	imagePullable = strings.TrimPrefix(imagePullable, "docker://")
+	imageComponents := strings.Split(imagePullable, "/")
+	imageName := imageComponents[len(imageComponents)-1]
+	imageRepo := strings.Join(imageComponents[:len(imageComponents)-1], "/")
+	// pkg:oci/adservice@sha256%3A45fb8ed886902c0c49e044b1f8870fad61c1022fa23c4943098302a8f1c5b75f?repository_url=gcr.io/google-samples/microservices-demo
+	imageField := fmt.Sprintf("pkg:oci/%s?repository_url=%s", url.PathEscape(imageName), url.PathEscape(imageRepo))
+	product := v1beta1.Product{
+		Component: v1beta1.Component{
+			ID: imageField,
+		},
+	}
+	product.Subcomponents = append(product.Subcomponents, v1beta1.Subcomponent{
+		Component: v1beta1.Component{
+			ID: packagePURL,
+		},
+	})
+	return &product, nil
+}
+
+func (a *APIServerStore) createVEX(ctx context.Context, cve domain.CVEManifest, cvep domain.CVEManifest, withRelevancy bool) error {
+	_, span := otel.Tracer("").Start(ctx, "APIServerStore.createVEX")
+	defer span.End()
+
+	imagePullable := cve.Annotations[v1.ImageIDMetadataKey]
+
+	// Timestamp
+	timestamp := time.Now().Format(time.RFC3339)
+
+	// Calculate VEX
+	vexDoc := v1beta1.VEX{
+		Metadata: v1beta1.Metadata{
+			Context:     "https://openvex.dev/ns/v0.2.0",
+			Author:      "kubescape.io",
+			AuthorRole:  "smart vulnerability scanner :-)",
+			Timestamp:   timestamp,
+			LastUpdated: timestamp,
+			Version:     0,
+			Tooling:     "kubescape-vulnerability-analyzer",
+		},
+	}
+
+	// Loop over the Vulnerability struct and add each vulnerability to the VEX document
+	for _, v := range cve.Content.Matches {
+		var aliases []string
+		for _, alias := range v.RelatedVulnerabilities {
+			aliases = append(aliases, string(alias.ID))
+		}
+
+		product, err := createProductStructForImageAndPackage(imagePullable, v.Artifact.PURL)
+
+		if err != nil {
+			return err
+		}
+
+		vexDoc.Statements = append(vexDoc.Statements, v1beta1.Statement{
+			Vulnerability: v1beta1.VexVulnerability{
+				ID:          v.Vulnerability.ID,
+				Name:        v.Vulnerability.DataSource,
+				Description: v.Vulnerability.Description,
+				Aliases:     aliases,
+			},
+
+			Products: []v1beta1.Product{
+				*product,
+			},
+
+			Status:          v1beta1.Status(vex.StatusNotAffected),
+			Justification:   v1beta1.Justification(vex.VulnerableCodeNotPresent),
+			ImpactStatement: "Vulnerable component is not loaded into the memory",
+		})
+	}
+
+	// Now change the status of the filtered vulnerabilities to "Affected"
+	for _, v := range cvep.Content.Matches {
+		for i, s := range vexDoc.Statements {
+			if s.Vulnerability.ID == v.Vulnerability.ID && v.Artifact.PURL == s.Products[0].Subcomponents[0].ID {
```

Comment:
Are you sure the lists are not empty? `s.Products[0].Subcomponents[0]`

Reply:
Right, I have consolidated this with a proper loop and tested